### PR TITLE
Update `@gear-js/data-storage` service in `idea`

### PR DESCRIPTION
**Pull request by Almazatun**
Resolves # .

- Update relationship between tables in `@gear-js/data-storage`.
- Update `@gear-js/api-gateway`
- Add the ability to handle Kafka `message` by `partition`  in `@gear-js/data-storage`. (depends on connected `node`)
- Combine `@gear-js/data-storage` with `@gear-js/event-listener`.
- Update `e2e` tests

@reviewer-or-team

### CODE REVIEW

**Review by osipov-mit — CHANGES_REQUESTED** [review]

File: idea/api-gateway/src/common/helpers/get-services-partition.ts [hunk 1]
```diff
@@ -0,0 +1,7 @@
+import { KAFKA_TOPICS } from '@gear-js/common';
+
+import { kafkaProducer } from '../../kafka/producer';
+
+export async function getServicesPartition(): Promise<void>{
+  await kafkaProducer.sendByTopic(KAFKA_TOPICS.SERVICES_PARTITION, '');
+}
```
[the remark this answers is not in the document]
Comment (Almazatun) [reply idea/api-gateway/src/common/helpers/get-services-partition.ts:7]: @osipov-mit The main idea of the function is to get a `Kafka partition` from running `@gear-js/data-storage` instances when the `api-gateway` service is restarted but the `data-storage` services are alive.

I will be glad to hear your recommendation to rename a meaningful `name`.

File: idea/data-storage/src/message/message.service.ts [hunk 2]
```diff
@@ -72,29 +56,17 @@ export class MessageService {
       }
 
       if (statuses[messageId] === 'Failed') {
-        const message = await this.messageRepository.get(messageId);
-        if (message.entry === 'Init') {
-          this.programService.setStatus(message.destination, genesis, InitStatus.FAILED);
+        const message = await this.messageRepository.get(messageId, genesis);
+        if (message.entry === MessageEntryPoint.INIT) {
+          await this.programService.setStatus(message.destination, genesis, ProgramStatus.INIT_FAILED);
```
Comment (osipov-mit) [inline idea/data-storage/src/message/message.service.ts:61]:
```suggestion
          await this.programService.setStatus(message.destination, genesis, ProgramStatus.TERMINATED);
```

**Review by osipov-mit — CHANGES_REQUESTED** [review]

File: idea/api-gateway/src/middleware/check-genesis.middleware.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+import { getResponse } from '../utils';
+
+export async function checkGenesisMiddleware(req: Request, res: Response, next: NextFunction) {
+  const body: IRpcRequest = req.body;
+
+  if(Array.isArray(body)) {
+    const isExistGenesis = body.every((value) => 'genesis' in value && value.genesis);
+
+    if (!isExistGenesis) {
+      res.send(getResponse(body, JSONRPC_ERRORS.Forbidden.name));
```
Comment (osipov-mit) [inline idea/api-gateway/src/middleware/check-genesis.middleware.ts:13]: Let's make this error more understandable

File: idea/api-gateway/src/routes/api-gateway/api-gateway.router.ts [hunk 4]
```diff
@@ -3,10 +3,11 @@ import { Request, Response, Router } from 'express';
 import { apiGatewayService } from '../../services/api-gateway.service';
 import { apiGatewayLogger } from '../../common/api-gateway.logger';
 import { testBalanceMiddleware } from '../../middleware';
+import { checkGenesisMiddleware } from '../../middleware';
 
 export const apiGatewayRouter = Router({});
 
-apiGatewayRouter.post('', testBalanceMiddleware, async (req: Request, res: Response) => {
+apiGatewayRouter.post('', testBalanceMiddleware, checkGenesisMiddleware, async (req: Request, res: Response) => {
```
Comment (osipov-mit) [inline idea/api-gateway/src/routes/api-gateway/api-gateway.router.ts:10]:
```suggestion
apiGatewayRouter.post('', checkGenesisMiddleware, testBalanceMiddleware, async (req: Request, res: Response) => {
```

File: idea/api-gateway/src/utils/get-response.ts [hunk 5]
```diff
@@ -1,17 +1,20 @@
...
+
+import { isJsonRpcAvailableMethod } from './is-json-rpc-available-method';
 
 export function getResponse(procedure: IRpcRequest, error?: any, result?: any): IRpcResponse {
   const response: IRpcResponse = {
     jsonrpc: '2.0',
     id: procedure.id,
   };
+
   if (error) {
```
Comment (osipov-mit) [inline idea/api-gateway/src/utils/get-response.ts:11]:
```suggestion
  if (error || result === undefined) {
```

File: idea/api-gateway/src/utils/get-response.ts [hunk 6]
```diff
@@ -1,17 +1,20 @@
...
     if (!JSONRPC_ERRORS[error]) {
       error = 'InternalError';
       console.log(error);
     }
     response['error'] = { message: JSONRPC_ERRORS[error].message, code: JSONRPC_ERRORS[error].code };
-  } else if (result) {
+  } else if (result || isJsonRpcAvailableMethod(procedure.method as API_METHODS)) {
```
Comment (osipov-mit) [inline idea/api-gateway/src/utils/get-response.ts:17]:
```suggestion
  } else {
```

File: idea/api-gateway/src/utils/is-json-rpc-available-method.ts [hunk 7]
```diff
@@ -0,0 +1,5 @@
+import { API_METHODS } from '@gear-js/common';
+
+export function isJsonRpcAvailableMethod(method: API_METHODS): boolean {
+  return [API_METHODS.TEST_BALANCE_AVAILABLE, API_METHODS.NETWORK_DATA_AVAILABLE].includes(method);
+}
```
Comment (osipov-mit) [inline idea/api-gateway/src/utils/is-json-rpc-available-method.ts:5]:
```suggestion
```

File: idea/common/src/interfaces/api-requests.ts [hunk 8]
```diff
@@ -17,9 +19,16 @@ interface GetOutgoingMessagesParams extends IGenesis, IPaginationParams, Pick<IM
 interface GetAllProgramsParams extends IGenesis, IPaginationParams, SearchParam {
   publicKeyRaw?: string;
   owner?: string;
+  fromDate?: string;
+  toDate?: string;
```
Comment (osipov-mit) [inline idea/common/src/interfaces/api-requests.ts:23]: I guess it's better to move it to a separate enum and extend the rest from it

**Review by osipov-mit — CHANGES_REQUESTED** [review]

File: idea/data-storage/src/common/enums/program-status.enum.ts [hunk 9]
```diff
@@ -0,0 +1,7 @@
+export enum ProgramStatus {
+  UNKNOWN = 'unknown',
+  ACTIVE = 'active',
+  TERMINATED = 'terminated',
+  INIT_FAILED = 'init_failed',
```
[the remark this answers is not in the document]
Comment (Almazatun) [reply idea/data-storage/src/common/enums/program-status.enum.ts]: Done

File: idea/common/src/interfaces/api-requests.ts (hunk 8 above)
Comment (Almazatun) [reply idea/common/src/interfaces/api-requests.ts:23]: Yeah, I think so

File: idea/api-gateway/src/middleware/check-genesis.middleware.ts (hunk 3 above)
Comment (Almazatun) [reply idea/api-gateway/src/middleware/check-genesis.middleware.ts:13]: @osipov-mit What would you suggest? To create a new `json-rpc` error or use an already existing `json-rpc` error
message like the bellow ?

 ```javascript 
  {
    name: 'InvalidParams',
    code: -32602,
    message: 'Invalid params',
  }
  ```

File: idea/data-storage/src/decorator/kafka-message-partition.decorator.ts [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+
+  descriptor.value = async function SafeWrapper() {
+    try {
+      const message = arguments[kafkaMessage] as Message;
+      if(message.partition === kafkaNetworkData.partition) return await originalMethod.apply(this, arguments);
+    } catch (ex) {
+      console.log(ex);
+      return { error: ex.name };
+    }
+  };
```
Comment (osipov-mit) [inline idea/data-storage/src/decorator/kafka-message-partition.decorator.ts:19]:
```suggestion
  descriptor.value = function SafeWrapper() {
    try {
      const message = arguments[kafkaMessage] as Message;
      if(message.partition === kafkaNetworkData.partition) return originalMethod.apply(this, arguments);
    } catch (ex) {
      console.log(ex);
      return Promise.resolve({ error: ex.name });
    }
  };
```

**Review by osipov-mit — CHANGES_REQUESTED** [review]

File: idea/data-storage/src/gear/gear-event-listener.ts [hunk 11]
```diff
@@ -0,0 +1,282 @@
...
+    });
+  }
+
+  private async handleEvents(method: string, payload: any): Promise<void> {
+    const { id, genesis, timestamp } = payload;
+
+    const eventsMethod = {
+      [Keys.UserMessageSent]: async () => {
+        const createMessageDBType = plainToClass(Message, {
+          ...payload,
```
Comment (osipov-mit) [inline idea/data-storage/src/gear/gear-event-listener.ts:113]: You need to find program here by source and place it to program field of the entity

File: idea/data-storage/src/gear/gear-event-listener.ts [hunk 12]
```diff
@@ -0,0 +1,277 @@
...
+        } = filterEvents(tx.hash, block, block.events, status).events.find(({ event }) =>
+          this.api.events.gear.MessageEnqueued.is(event),
+        ).event as MessageEnqueued;
+
+        let codeId;
+        const uploadProgramIndex = 0;
+        const createProgramIndex = 1;
+
+        if(tx.method.method === txMethods[uploadProgramIndex]) codeId = generateCodeHash(tx.args[0].toHex());
+        if(tx.method.method === txMethods[createProgramIndex]) codeId = tx.args[0].toHex();
```
Comment (osipov-mit) [inline idea/data-storage/src/gear/gear-event-listener.ts:215]:
```suggestion
    let codeId = tx.method.method === 'uploadProgram' ? generateCodeHash(tx.args[0].toHex()) : tx.args[0].toHex();
```

File: idea/data-storage/src/gear/gear-event-listener.ts [hunk 13]
```diff
@@ -0,0 +1,277 @@
...
+          id: destination.toHex(),
+          blockHash: block.createdAtHash.toHex(),
+          timestamp,
+          code: await this.codeRepository.get(codeId, this.genesis),
+          genesis: this.genesis,
+        });
+      }
+    }
+
+    if (programs.length >= 1) await this.programService.createPrograms(programs);
```
Comment (osipov-mit) [inline idea/data-storage/src/gear/gear-event-listener.ts:228]: Is there a case where `programs.length` can be 0? It's already checked that `extrinsic.length` > 0, in this case you can be sure that programs won't be equal 0. But you need to move it to the previous block

**Review by osipov-mit — CHANGES_REQUESTED** [review]

File: idea/data-storage/src/gear/gear-event-listener.ts (hunk 13 above)
Comment (Almazatun) [reply idea/data-storage/src/gear/gear-event-listener.ts:228]: I get what you mean. Will update all three handlers.

**Review by sergeyfilyanin — APPROVED** [review]

**Pull request merged**